File: SECD/SECD.hs
module SECD where

import Syntax
import Instructions
import Data.List (elemIndex)
import Data.Map (Map)
import qualified Data.Map as Map
import Parse

type Addr = Int
type Stack = [ValueSECD]
type EnvSECD = [ValueSECD]
type Dump = [(Stack,EnvSECD,Code)]
type Symtable = [Id]
type Closure = (Code, EnvSECD)
type Memory = Map Addr Closure
type SECD = (Stack,EnvSECD,Code,Dump,Memory)

data ValueSECD = Prim Int 
 | Address Addr
 deriving(Show)

compiler::String -> Code
compiler s = compile (Parse.parse s) []

compile:: Term -> Symtable -> Code
compile (Const c) sym = [LDC c]
compile (Var i) sym = --[LD (elemIndexInt i sym)]
	case elemIndex i sym of
 	 Just v -> [LD v]
 	 Nothing -> error "variavel livre"

compile (Lambda i t) sym= [LDF ((compile t sym')++[RTN])]
	where sym' = i:sym
compile (App t1 t2) sym = (compile t1 sym) ++ (compile t2 sym) ++ [AP]
compile (t1 :+ t2) sym= (compile t1 sym) ++ (compile t2 sym) ++ [ADD]
compile (t1 :- t2) sym= (compile t1 sym) ++ (compile t2 sym) ++ [SUB]
compile (t1 :* t2) sym= (compile t1 sym) ++ (compile t2 sym) ++ [MUL]
compile (Ifzero t1 t2 t3) sym= (compile t1 sym) ++ [SEL c1 c2]
	where 
		c1 = ((compile t2 sym) ++ [JOIN])
		c2 = ((compile t3 sym) ++ [JOIN])
compile (Let i t1 t2) sym= compile (App (Lambda i t2) t1) sym
compile (Fix (Lambda f (Lambda x e))) sym = [LDRF (compile e sym' ++ [RTN])]
	where sym' = x:f:sym

--call by value
execute::SECD -> SECD
execute (s,e,(LDC i):c,d,m) = (((Prim i):s),e,c,d,m)
--execute (s,e, (LD (-1)):c,d,m) = (s,e,c,d,m)
execute (s,e, (LD i):c,d,m) = (v:s,e,c,d,m)
 where v = lookupId i e 
execute (s,e,(LDF code):c,d,m) = 
 let a = next m
     store' = Map.insert a (code,e) m
 in ((Address a):s,e,c,d,store')
execute (s,e,(LDRF c'):c,d,m) = ((Address a):s,e,c,d,m')
 where a = next m
       m' =(Map.insert a (c',(Address a:e)) m)
execute (Prim v1:Prim v2:s,e, (ADD):c,d,m) = (Prim (v1+v2):s,e,c,d,m)
execute (Prim v1:Prim v2:s,e, (SUB):c,d,m) = (Prim (v2-v1):s,e,c,d,m)
execute (Prim v1:Prim v2:s,e, (MUL):c,d,m) = (Prim (v1*v2):s,e,c,d,m)
execute (v:(Address a):s,e,AP:c,d,m) = ([],v:e',c',(s,e,c):d,m)
 where (c',e')= 
 	case (Map.lookup a m) of
 		Just (c',e') -> (c',e')
 		Nothing -> error "closure "
execute (v:s,e,RTN:c,(s',e',c'):d,m) = (v:s',e',c',d,m)
execute ((Prim 0):s,e,(SEL c1 c2):c,d,m) = (s,e,c1,([],[],c):d,m)
execute ((Prim _):s,e,(SEL c1 c2):c,d,m) = (s,e,c2,([],[],c):d,m)
execute (s,e,JOIN:c,(_,_,c'):d,m) = (s,e,c',d,m)
execute (s,e,HALT:c,d,m) = (s,e,[],d,m)
execute secd = error "indefinido"

{-elemIndex::String -> Symtable -> Int
elemIndex s [] = 0
elemIndex s (x:xs) 
 | s == x = 0 
 | otherwise = 1 + (elemIndex s xs)-}

elemIndexInt::Eq a => a -> [a] -> Int 
elemIndexInt n l = unpack $ elemIndex n l

unpack:: Maybe Int -> Int
unpack (Just a) = a
unpack Nothing = -1

lookupId::Int->EnvSECD-> ValueSECD
lookupId i env = env!!i

next::Memory -> Addr
next store = 1 + (Map.size store)

run::Code -> IO()
run code = sequence_ [print state | state <- trace]
 where states = iterate execute ([],[],code++[HALT],[],Map.empty)
       trace = takeWhile (not.final) states
       final (s,e,c,d,m) = null c

--exemplos
ex1= (compiler "(\\x.x)(2)")
ex2= [LDC 1, LDC 2, LDC 3, MUL, ADD]
ex3= [LDF [LD 0, LDC 1, ADD, RTN]]
ex4= (compiler "(\\x.ifzero x 1 x+1)(1)")
ex5= [LDF [LD 0, SEL [LDC 1, JOIN] [LDC 0, JOIN], RTN]]
ex6= (compiler "let x = 1 in x+2")
ex7= (compiler "(\\x.(\\y.ifzero x+y 1 x+y)(0))(2)")
--está mal (entra em ciclo)
ex8= (compile (Let "fact" (Fix (Lambda "f" (Lambda "n" (Ifzero (Var "n") (Const 1) ((Var "n"):*(App (Var "f") ((Var "n"):-(Const 1)))))))) (App (Var "fact") (Const 3))) [])

File: SECD/SECD.c
#include<stdio.h>
#include<inttypes.h>
#include<stdlib.h>
#include <assert.h>

#include "SECD.h"

int code[] = {1,1,1,2,1,3,5,5,0};
value_t stack[];
dump_t dump [];

env_t extend(value_t elem,env_t env){
	env_t new= (env_t)malloc(sizeof(env_t));
	assert (new!=NULL);
	new->val = elem;
	new->next=env;
	return new;
}

value_t lookup(int n, env_t env){
	printf("Está no lookup, n: %d",n);
	env_t new_env;
	new_env=env;
	for(int i=0;i<n;i++){
		new_env=new_env->next;
	}
	return new_env->val;
}

closure_t* make_closure(int pc, env_t env){
	closure_t *new=(closure_t*)malloc(sizeof(closure_t));
	new->pc=pc;
	new->env=env;
	return new;
}

value_t run(void){
	int pc=0;
	int sp=0;
	int dp=0;
	env_t env = NULL;

	for(;;){
		value_t opa, opb;
		int t;
		closure_t *cptr;
		env_t nenv;
		int opcode=code[pc];
		switch(opcode){
			case LDC:
				printf("está no LDC\n");
				opa=(value_t)code[++pc];
				printf("opa: %" PRIxPTR "\n", opa);
				stack[sp]=opa;
				printf("sp: %d \n",sp);
				printf("stack[sp]: %" PRIxPTR "\n", stack[sp]);
				printf("pc: %d \n",pc);
				sp++;
				pc++;
				printf("sp: %d \n",sp);
				printf("pc: %d \n",pc);
				break;
			case LD:
				printf("está no LD\n");
				t=code[pc++];
				printf("t: %d \n",t);
				stack[sp] = lookup(t,env);
				sp++;
				break;
			case LDF:
				t=code[pc++];
				cptr = make_closure(t,env);
				stack[sp++]=(value_t)cptr;
				break;
			case LDRF:
				t=code[pc++];
				nenv=extend((value_t)NULL,env);
				cptr=make_closure(t,nenv);
			case ADD:
				printf("está no ADD\n");
				opa=stack[--sp];
				opb=stack[--sp];
				printf("opa: %" PRIxPTR " opb: %" PRIxPTR "\n", opa, opb);
				stack[sp++]=opa+opb;
				int j;
				for(j = 0; stack[j] != '\0'; j++)
			    	printf("stack[j]: %" PRIxPTR "\n",stack[j]);
				pc++;
				break;
			case SEL:
				opa=stack[--sp];
				pc=pc+2;
				dp++;
				if(opa==0){
					pc=code[pc];
				}
				else{
					pc=code[pc++];
				}
				break;
			case JOIN:
				pc=dump[dp].pc;
				break;
			case RTN:
				pc=dump[dp].pc;
				env=dump[dp].env;	
				break;
			case AP:
				opa=stack[--sp];
				value_t vl=stack[--sp];
				cptr=(closure_t*)vl;
				dump[dp].pc=pc;
				dump[dp].env=env;
				dp++;
				env=extend(opa,cptr->env);
				pc=cptr->pc;
				break;
			case Halt:
				printf("Entrou no halt\n");
				//size_t n= sizeof(stack)/sizeof(stack[0]); 
			    return stack[0];
				break;
		}
	}
	
}

int main(){
	value_t val;
	val = run();
	printf("Resultado: %" PRIxPTR "\n", val);
}